dao/tenant_dao.go

```diff
@@ -70,3 +77,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
 
 	return &tenant, nil
 }
+
+func (t *tenantDaoImpl) GetUntranslatedTenants() ([]m.Tenant, error) {
+	var tenants []m.Tenant
+
+	err := DB.Debug().
+		Model(&m.Tenant{}).
+		Where(`("external_tenant" IS NOT NULL OR "external_tenant" != '') AND ("org_id" IS NULL OR "org_id" = '')`).
+		Find(&tenants).
+		Error
+
+	if err != nil {
+		return nil, err
+	}
+
+	return tenants, err
+}
+
+func (t *tenantDaoImpl) TranslateTenants() (int64, uint64, uint64, []m.TenantTranslation, error) {
+	// The where condition to fetch only the untranslated tenants.
+	untranslatedTenantsWhereCondition := `("external_tenant" IS NOT NULL OR "external_tenant" != '') AND ("org_id" IS NULL OR "org_id" = '')`
+
+	// Count the total number of translatable tenants.
+	var translatableTenants int64
+	// Count the translation operations.
+	var translatedTenants, untranslatedTenants uint64
+	// Keep track of each of the tenant translations.
+	var tenantTranslations []m.TenantTranslation
+
+	// Run everything inside a transaction to avoid inconsistencies. This way if new tenants are added/removed while
+	// the translation is in progress, all the counts and the results will still match.
+	err := DB.Debug().Transaction(func(tx *gorm.DB) error {
+		// Get the count to be able to process the tenants in batches of 100 tenants.
+		err := tx.
+			Model(&m.Tenant{}).
+			Where(untranslatedTenantsWhereCondition).
+			Count(&translatableTenants).
+			Error
+
+		if err != nil {
+			return fmt.Errorf("unable to count untranslated tenants: %w", err)
+		}
+
+		translator := tenantid.NewTranslator(config.Get().TenantTranslatorUrl)
+		for i := int64(0); i < translatableTenants; i += 100 {
+			// Grab all the EBS account numbers to be processed.
+			var ebsAccountNumbers []string
+			err := tx.
+				Model(&m.Tenant{}).
+				Where(untranslatedTenantsWhereCondition).
+				Pluck("external_tenant", &ebsAccountNumbers).
+				Offset(int(i)).
+				Limit(100).
```

Comment:
I guess we need to add here Order BY clause as order of records is not ensured.

Reply:
Oh but maybe it will not causes any issues as there is `Where(untranslatedTenantsWhereCondition)`, I am taking it back.

Reply:
Mmmmmmmm... I don't follow. Why do we want ordered records here?

Reply:
Idea is when order of records is not ensured then it can take first 100 records(for offset=0) and then it can pick also some same records for next 100 records - 

is it probably edge case and most likely it will not cause issue - tenants from batch are processed and org_id is set and then those tenants are excluded from this query thanks to the  `untranslatedTenantsWhereCondition`
- so it pertains just records which remain after translation processing without `org_id`.

hopefully it makes sense.

Thanks for adding order by!

Reply:
Thanks an interesting point, yeah. Thanks for the explanation! :+1: